src/constructs/lambda/event-sources.ts

```diff
@@ -0,0 +1,93 @@
+import type { IEventSourceDlq } from "@aws-cdk/aws-lambda";
+import type { KinesisEventSourceProps } from "@aws-cdk/aws-lambda-event-sources";
+import { Duration } from "@aws-cdk/core";
+
+export type StreamProcessingProps = Omit<
+  KinesisEventSourceProps,
+  "bisectBatchOnError" | "maxRecordAge" | "onFailure" | "retryAttempts"
+>;
+
+/**
+ * Use with caution! This will cause your lambda to continuously retry if it encounters errors
+ * whilst processing a batch. If the errors are permanent (e.g. due to a bad record) your lambda
+ * will fall behind with stream processing i.e. it will only move onto new batches when the problematic
+ * records expire. Consider using [[`StreamErrorHandlingProps`]] instead.
+ *
+ * When using this approach to error handling, ensure that an alarm is configured to monitor
+ * lambda failures.
+ */
+export interface BlockProcessingAndRetryIndefinitely {
+  blockProcessingAndRetryIndefinitely: true;
+}
+
+/**
+ * In order to prevent your lambda from continuously retrying if it encounters errors
+ * whilst processing a batch, use `retryBehaviour` to give up on a record once it reaches a certain age,
+ * or after a specified number of attempts. See [[`StreamRetry`]] for more details.
+ *
+ * In order to isolate bad records as part of the retry process, you may also want to bisect problematic batches
+ * using `bisectBatchOnError`. For example:
+ *
+ * ```typescript
+ * const errorHandlingProps: ErrorHandlingProps = {
+ *   bisectBatchOnError: true,
+ *   retryBehaviour: StreamRetry.maxAttempts(5),
+ * }
+ * ```
+ *
+ * Records which could not be processed successfully can (optionally) be sent to a dead letter queue
+ * via `deadLetterQueueForSkippedRecords`.
+ */
+export interface StreamErrorHandlingProps {
+  retryBehaviour: StreamRetry;
+  bisectBatchOnError: boolean;
+  deadLetterQueueForSkippedRecords?: IEventSourceDlq;
```

Comment:
We could think about providing a queue for users if they select this option, but this PR is already getting quite big!